### ISSUE TITLE
Multiple issues with emoji matching

### ISSUE THREAD

**Issue by mosra** [doxygen/doxygen#6691]
Similar to #6612, related to emoji support (#6517) added in #6534 by @albert-github. I have multiple false positives on emoji being detected in [Magnum](https://magnum.graphics) documentation. One example is this:

```cpp

/**
…

Aspect ratio    | CSS class (landscape/portrait)
--------------- | ------------------------------
1:1             | `.aspect-1-1` 
4:3 (1.33:1)    | `.aspect-4-3` (default) / `.aspect-3-4`
3:2 (1.5:1)     | `.aspect-3-2` / `.aspect-2-3`
16:9 (1.78:1)   | `.aspect-16-9` / `.aspect-9-16`
2:1             | `.aspect-2-1` / `.aspect-1-2`

…
*/
```

producing absolutely unwanted and totally bogus warnings such as:

```
platforms-html5.dox:293: warning: Unsupported emoji ':3 (1.33:' found
platforms-html5.dox:297: warning: Unsupported emoji ':2 (1.5:' found
platforms-html5.dox:301: warning: Unsupported emoji ':9 (1.78:' found
```

Another is in code snippets like

```cpp
/**
… `Platform::*Application::viewportEvent(const Vector2i&)` …
*/
```

where the asterisk is a wildcard for multiple types such as `Sdl2Application`, `GlfwApplication` etc., producing 

```
changelog.dox:661: warning: Unsupported emoji ':*Application:' found
```

Some suggestions about how to fix this:

- match emoji names as `[a-zA-Z0-9_-]` and nothing else (I don't think I ever encountered any other characters)
- don't accept a space in a emoji name (*seriously?!*)
- don't *ever* look for emojis in code snippets and inline code (I doubt there's any use case for that, if really needed, people can always insert the matching unicode character directly)
- have emojis **opt-in** via a config option, because adding `\:` everywhere just because there's a new feature with dubious use is absolutely *not* acceptable

**Comment by albert-github** [comment]
Some thoughts about the points mentioned:
- match emoji names as [a-zA-Z0-9_-] and nothing else (I don't think I ever encountered any other characters)
  When looking at the names of the as defined in https://unicode.org/emoji/charts/full-emoji-list.html we see here quite a few "CLDR short names" are with special characters e.g.:
  - `keycap: *`
  - `+1`
  - ` b button (blood type)`
  - `flag Antigua & Barbuda`
  - `mrs. claus`
- don't accept a space in a emoji name (seriously?!)
  I don't get this point, spaces are possible just not directly after or before the `:` when handling the emoji.
- don't ever look for emojis in code snippets and inline code (I doubt there's any use case for that, if really needed, people can always insert the matching unicode character directly)
  probably there is no use of emoji in code snippets etc. but it quite difficult to filter them out.
- have emojis opt-in via a config option, because adding \: everywhere just because there's a new feature with dubious use is absolutely not acceptable
  I do agree that in case not found emoji it is a bit hard (understatement) to have to specify `\:` everywhere. I can think of multiple solutions here:
  - conditionally enable emoji (as suggested in the remark)
  - make a conditional warning out of it.

Some points I have to look at:
- By head I thought that in case of a not recognized emoji the text was printed just as is but I see that this is not the case.
- I thought that I fixed the problems with `::text::`  as not possible as an emoji (as you mentioned #6612 / #6614) . but apparently we have here a new one.

**Comment by doxygen** [comment]
I'm also worried about false detections of emoji's in legacy documentation. 

I think it could be a solution to limit ourselves to the GitHub defined emojis (i.e. https://api.github.com/emojis). That avoids the 'spaces and strange characters' in emojis and the false detections that can be the result of it. 

Maybe we should add special patterns for `:+1:`, `:-1:`, so that we do not need to include + and - in the pattern and next to that only the pattern ":"[a-z_0-9]*":" should be considered.

Next to that of course also issues with `::text::` should fixed and not recognized emojis should pass to the output untouched.

**Comment by albert-github** [comment]
Regarding the points:
- not recognized emojis should pass to the output untouched, I've just pushed a proposed patch, pull request #6698 
- issues with `::text::`, I've just pushed a proposed patch, pull request #6699.
  An emoji pattern is not seen as an emoji anymore in case it is directly followed by `:`.

**Comment by mosra** [comment]
Let's step back a bit:

- Do you seriously think someone would hand-enter `:japanese "congratulations" button:` for an emoji name? Or any other "unicode short emoji name" containing spaces? I doubt that.
- There's a reason why the list of short names introduced by GitHub and adopted by GitLab, Gitter, Slack and others has the name character set restricted to `[a-zA-Z0-9_+-]`. Because everything else has way too many false positives. Again, **especially spaces**.
- What's more likely when I want to insert a random emoji to the text?
    a. I search for it on google or use an emoji keyboard on my system and copy-paste the Unicode representation of it
    b. I search for it on google and instead of copy-pasting the Unicode representation that's [directly on the results page](https://www.google.com/search?q=congratulations%20emoji) I spend extra time looking up the canonical "unicode short emoji name" so I can then copy the text, wrap it in `:` and then hope the parser catches it and parses correctly?

> probably there is no use of emoji in code snippets etc. but it quite difficult to filter them out.

Seriously? There's already code in place that doesn't convert HTML tags that are inside code snippets. So I guess similar logic could be reused for emoji as well. 

There are also markup formats such as reStructuredText (used by [Breathe](https://breathe.readthedocs.io/en/latest/) to add arbitrary contents to Doxygen code blocks), which looks like this -- note that basically every occurence of its text role command would qualify as emoji -- and you *don't want* these to be converted to emojis:

```rst
…

-   To maintain accessibility easier, makes it possible to enforce :rst:`:alt:`
    text for every image and figure by setting :py:`M_IMAGES_REQUIRE_ALT_TEXT`
    to :py:`True`.

You can add `additional CSS classes <{filename}/css/components.rst#images>`_ to
images or figures via the :rst:`:class:` or :rst:`:figclass:` options,
respectively. If you want the image or figure to be clickable, use the
:rst:`:target:` option. The alt text can be specified using the :rst:`:alt:`
option for both images and figures.

…
```

No matter the legacy-ness of the existing documentation, from now on after every modification I have to go through all my output and check it for presence of rogue emoji. Not to mention that in some contexts and use cases having emoji present in the output would be *absolutely unacceptable*. So what I propose is:

- restrict the "Markdown emoji" to just what GitHub Flavored Markdown supports and *not more*, ideally under an *opt-in* config option so tools like Breathe continue to work properly
- and, for inserting random other emoji, introduce a new Doxygen-specific command `@emoji`, that accepts the long-name. Or, *better*, introduce a new command `@unicode`, accepting (a sequence of) UTF-32 codepoints, so people can use it not only for emoji but also for much more useful things like adding `×`, `÷` and other characters that are not readily available on most keyboards.

**Comment by albert-github** [comment]
Issue #6698 and #6699 have been integrated on github

**Comment by albert-github** [comment]
@mosra and how would you handle output formats like LaTeX?

**Comment by doxygen** [comment]
Based on your feedback and concerns I've decided to change the emoji implementation.

- It now requires an explicit `\emoji` command instead of relying on `:name:` patterns
- The list of supported emoji characters is limited to that of GitHub (so without the spaces). I've included a script to download the GitHub emoji images (and provided a download link with a zip as well) to support LaTeX output.
- The download script also allows recreation of the github_name to unicode mapping table inside doxygen.
- In the XML an emoji character gets its own &lt;emoji/&gt; tag with name and unicode attributes.
- Documentation has been updated accordingly.

Commit https://github.com/doxygen/doxygen/commit/c3ee766d0ad5721c753581e7f87026614c0730e1

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.15.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).

**Comment by mosra** [comment]
Thank you, this change makes it much more robust.
